// ngx-grid-stack/esm2015/lib/grid-stack/grid-stack.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, QueryList, Input, ContentChildren, ElementRef, Renderer } from '@angular/core';
import { GridStackOptions } from './../models/grid-stack-options.model';
import { GridStackItemComponent } from './../grid-stack-item/grid-stack-item.component';
export class GridStackComponent {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.options = new GridStackOptions();
        this.gridStack = null;
        this.grid = null;
        this.defaultOptions = {
            cellHeight: '60px',
            width: 12,
            height: 0,
            animate: true,
            float: false,
            resizable: true
        };
    }
    /**
     * @param {?} item
     * @return {?}
     */
    makeWidget(item) {
        item.jGridRef = this.grid;
        if (item.option != null && item.option.noResize != null && item.option.noResize) {
            return;
        }
        this.grid.resizable(item.nativeElement, true);
        this.grid.move(item.nativeElement, item.option.x, item.option.y);
        this.grid.resize(item.nativeElement, item.option.width, item.option.height);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    updateWidget(item) {
        this.grid.resizable(item.nativeElement, true);
        this.grid.move(item.nativeElement, item.option.x, item.option.y);
        this.grid.resize(item.nativeElement, item.option.width, item.option.height);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    AddWidget(item) {
        item.jGridRef = this.grid;
        if (item.option != null && item.option.noResize != null && item.option.noResize) {
            return;
        }
        this.grid.makeWidget(item.nativeElement);
        this.grid.resizable(item.nativeElement, true);
        this.grid.move(item.nativeElement, item.option.x, item.option.y);
        this.grid.resize(item.nativeElement, item.option.width, item.option.height);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    RemoveWidget(item) {
        this.grid.removeWidget(item.nativeElement, false);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        const /** @type {?} */ that = this;
        const /** @type {?} */ nativeElement = this.el.nativeElement;
        if (this.options == null) {
            this.options = new GridStackOptions();
        }
        for (const /** @type {?} */ key of Object.keys(this.defaultOptions)) {
            if (!this.options.hasOwnProperty(key)) {
                this.options = this.defaultOptions[key];
            }
        }
        this.renderer.setElementAttribute(nativeElement, 'data-gs-width', String(this.options.width));
        this.renderer.setElementAttribute(nativeElement, 'data-gs-height', String(this.options.height));
        this.gridStack = jQuery(nativeElement).gridstack(this.options);
        this.grid = this.gridStack.data('gridstack');
        this.gridStack.on('change', (e, items) => {
            _.each(items, (item) => this.widgetChanged(item));
        });
        // Initialize widgets
        this.items.forEach(item => that.makeWidget(item));
    }
    /**
     * @param {?} change
     * @return {?}
     */
    widgetChanged(change) {
        const /** @type {?} */ jWidget = change.el;
        const /** @type {?} */ gridStackItem = this.items.find(item => item.jWidgetRef !== null ? item.jWidgetRef === jWidget[0] : false);
        if (!gridStackItem) {
            return;
        }
        gridStackItem.update(change.x, change.y, change.width, change.height);
    }
}
GridStackComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-grid-stack',
                template: `<ng-content></ng-content>
`,
                styles: [':host { display: block; }']
            },] },
];
/** @nocollapse */
GridStackComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
GridStackComponent.propDecorators = {
    options: [{ type: Input }],
    items: [{ type: ContentChildren, args: [GridStackItemComponent,] }]
};
function GridStackComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridStackComponent.prototype.options;
    /** @type {?} */
    GridStackComponent.prototype.items;
    /** @type {?} */
    GridStackComponent.prototype.gridStack;
    /** @type {?} */
    GridStackComponent.prototype.grid;
    /** @type {?} */
    GridStackComponent.prototype.defaultOptions;
    /** @type {?} */
    GridStackComponent.prototype.el;
    /** @type {?} */
    GridStackComponent.prototype.renderer;
}
