// ngx-grid-stack/esm2015/lib/grid-stack-item/grid-stack-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, ElementRef, ViewChild, Renderer, EventEmitter, ViewContainerRef } from '@angular/core';
import { GridStackItem } from './../models/grid-stack-item.model';
export class GridStackItemComponent {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.GridConfigurationChanged = new EventEmitter();
        this.contentComponentRef = null;
        this.jGridRef = null;
        this.jWidgetRef = null;
        this.jWidgetRef = el.nativeElement;
    }
    /**
     * @return {?}
     */
    get nativeElement() {
        return this.el.nativeElement;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.RenderWidget(null);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    UpdateWidget(item) {
    }
    /**
     * @param {?} item
     * @return {?}
     */
    RenderWidget(item) {
        const /** @type {?} */ renderer = this.renderer;
        if (item != null) {
            this.option = item;
        }
        this.renderer.setElementAttribute(this.nativeElement, 'style', 'margin-left:' + this.option.marginWidth + ';');
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-x', String(this.option.x));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-y', String(this.option.y));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-width', String(this.option.width));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-height', String(this.option.height));
        if (this.option.minWidth) {
            renderer.setElementAttribute(this.nativeElement, 'data-gs-min-width', String(this.option.minWidth));
        }
        if (this.option.noResize != null && this.option.noResize === true) {
            renderer.setElementAttribute(this.nativeElement, 'data-gs-no-resize', 'yes');
        }
    }
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    update(x, y, width, height) {
        // console.log('here');
        if (x === this.option.x && y === this.option.y && width === this.option.width && height === this.option.height) {
            return;
        }
        if (this.option != null) {
            this.option.x = x;
            this.option.y = y;
            this.option.width = width;
            this.option.height = height;
            const /** @type {?} */ optionNew = GridStackItem.Clone(this.option);
            this.GridConfigurationChanged.emit(optionNew);
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.contentComponentRef !== null) {
            this.contentComponentRef.destroy();
        }
    }
}
GridStackItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-grid-stack-item',
                template: `<div class="grid-stack-item-content">
  <div #contentPlaceholder *ngIf="contentTemplate"></div>
  <ng-content *ngIf="!contentTemplate"></ng-content>
</div>
`,
                styles: [``]
            },] },
];
/** @nocollapse */
GridStackItemComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer }
];
GridStackItemComponent.propDecorators = {
    contentPlaceholder: [{ type: ViewChild, args: ['contentPlaceholder', { read: ViewContainerRef },] }],
    contentTemplate: [{ type: Input }],
    option: [{ type: Input }],
    GridConfigurationChanged: [{ type: Output }]
};
function GridStackItemComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridStackItemComponent.prototype.contentPlaceholder;
    /** @type {?} */
    GridStackItemComponent.prototype.contentTemplate;
    /** @type {?} */
    GridStackItemComponent.prototype.option;
    /** @type {?} */
    GridStackItemComponent.prototype.GridConfigurationChanged;
    /** @type {?} */
    GridStackItemComponent.prototype.contentComponentRef;
    /** @type {?} */
    GridStackItemComponent.prototype.jGridRef;
    /** @type {?} */
    GridStackItemComponent.prototype.jWidgetRef;
    /** @type {?} */
    GridStackItemComponent.prototype.el;
    /** @type {?} */
    GridStackItemComponent.prototype.renderer;
}
