// ngx-grid-stack/esm5/lib/grid-stack-item/grid-stack-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, ElementRef, ViewChild, Renderer, EventEmitter, ViewContainerRef } from '@angular/core';
import { GridStackItem } from './../models/grid-stack-item.model';
var GridStackItemComponent = /** @class */ (function () {
    function GridStackItemComponent(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.GridConfigurationChanged = new EventEmitter();
        this.contentComponentRef = null;
        this.jGridRef = null;
        this.jWidgetRef = null;
        this.jWidgetRef = el.nativeElement;
    }
    Object.defineProperty(GridStackItemComponent.prototype, "nativeElement", {
        get: /**
         * @return {?}
         */
        function () {
            return this.el.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    GridStackItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.RenderWidget(null);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    GridStackItemComponent.prototype.UpdateWidget = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
    };
    /**
     * @param {?} item
     * @return {?}
     */
    GridStackItemComponent.prototype.RenderWidget = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var /** @type {?} */ renderer = this.renderer;
        if (item != null) {
            this.option = item;
        }
        this.renderer.setElementAttribute(this.nativeElement, 'style', 'margin-left:' + this.option.marginWidth + ';');
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-x', String(this.option.x));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-y', String(this.option.y));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-width', String(this.option.width));
        this.renderer.setElementAttribute(this.nativeElement, 'data-gs-height', String(this.option.height));
        if (this.option.minWidth) {
            renderer.setElementAttribute(this.nativeElement, 'data-gs-min-width', String(this.option.minWidth));
        }
        if (this.option.noResize != null && this.option.noResize === true) {
            renderer.setElementAttribute(this.nativeElement, 'data-gs-no-resize', 'yes');
        }
    };
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    GridStackItemComponent.prototype.update = /**
     * @param {?} x
     * @param {?} y
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function (x, y, width, height) {
        // console.log('here');
        if (x === this.option.x && y === this.option.y && width === this.option.width && height === this.option.height) {
            return;
        }
        if (this.option != null) {
            this.option.x = x;
            this.option.y = y;
            this.option.width = width;
            this.option.height = height;
            var /** @type {?} */ optionNew = GridStackItem.Clone(this.option);
            this.GridConfigurationChanged.emit(optionNew);
        }
    };
    /**
     * @return {?}
     */
    GridStackItemComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    GridStackItemComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.contentComponentRef !== null) {
            this.contentComponentRef.destroy();
        }
    };
    GridStackItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-grid-stack-item',
                    template: "<div class=\"grid-stack-item-content\">\n  <div #contentPlaceholder *ngIf=\"contentTemplate\"></div>\n  <ng-content *ngIf=\"!contentTemplate\"></ng-content>\n</div>\n",
                    styles: [""]
                },] },
    ];
    /** @nocollapse */
    GridStackItemComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer }
    ]; };
    GridStackItemComponent.propDecorators = {
        contentPlaceholder: [{ type: ViewChild, args: ['contentPlaceholder', { read: ViewContainerRef },] }],
        contentTemplate: [{ type: Input }],
        option: [{ type: Input }],
        GridConfigurationChanged: [{ type: Output }]
    };
    return GridStackItemComponent;
}());
export { GridStackItemComponent };
function GridStackItemComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    GridStackItemComponent.prototype.contentPlaceholder;
    /** @type {?} */
    GridStackItemComponent.prototype.contentTemplate;
    /** @type {?} */
    GridStackItemComponent.prototype.option;
    /** @type {?} */
    GridStackItemComponent.prototype.GridConfigurationChanged;
    /** @type {?} */
    GridStackItemComponent.prototype.contentComponentRef;
    /** @type {?} */
    GridStackItemComponent.prototype.jGridRef;
    /** @type {?} */
    GridStackItemComponent.prototype.jWidgetRef;
    /** @type {?} */
    GridStackItemComponent.prototype.el;
    /** @type {?} */
    GridStackItemComponent.prototype.renderer;
}
