// ngx-grid-stack/esm5/lib/models/grid-stack-item.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var GridStackItem = /** @class */ (function () {
    function GridStackItem() {
        this.x = 0;
        this.y = 0;
        this.height = 1;
        this.width = 1;
        this.noResize = false;
        this.autoPosition = false;
        this.marginWidth = '10px';
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    GridStackItem.Clone = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        var /** @type {?} */ result = new GridStackItem();
        result.autoPosition = widget.autoPosition;
        result.customId = widget.customId;
        result.el = widget.el;
        result.height = widget.height;
        result.locked = widget.locked;
        result.maxHeight = widget.maxHeight;
        result.maxWidth = widget.maxWidth;
        result.minHeight = widget.minHeight;
        result.minWidth = widget.minWidth;
        result.noMove = widget.noMove;
        result.noResize = widget.noResize;
        result.width = widget.width;
        result.x = widget.x;
        result.y = widget.y;
        return result;
    };
    return GridStackItem;
}());
export { GridStackItem };
function GridStackItem_tsickle_Closure_declarations() {
    /** @type {?} */
    GridStackItem.prototype.x;
    /** @type {?} */
    GridStackItem.prototype.y;
    /** @type {?} */
    GridStackItem.prototype.height;
    /** @type {?} */
    GridStackItem.prototype.width;
    /** @type {?} */
    GridStackItem.prototype.maxHeight;
    /** @type {?} */
    GridStackItem.prototype.minHeight;
    /** @type {?} */
    GridStackItem.prototype.maxWidth;
    /** @type {?} */
    GridStackItem.prototype.minWidth;
    /** @type {?} */
    GridStackItem.prototype.noResize;
    /** @type {?} */
    GridStackItem.prototype.noMove;
    /** @type {?} */
    GridStackItem.prototype.autoPosition;
    /** @type {?} */
    GridStackItem.prototype.marginWidth;
    /** @type {?} */
    GridStackItem.prototype.locked;
    /** @type {?} */
    GridStackItem.prototype.el;
    /** @type {?} */
    GridStackItem.prototype.customId;
}
